### ISSUE TITLE
Variable type inference doesn't handle loops

### ISSUE THREAD

**Issue by eltoder** [python/mypy#15982]
Apologies if this was already reported. I could not find an existing issue.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.11&gist=7b4923efe4d5bad055733ce5e8f9a2f7
```python
def fn(x: int | None) -> None:
    pass

def test(n: int) -> None:
    x = None
    for _ in range(n):
        fn(x)
        x = "123"
```

**Expected Behavior**

This should give an error because the type of `x` in `test` is `str | None`, but `fn` requires `int | None`.

**Actual Behavior**

No errors.

**Your Environment**

- Mypy version used: 1.5.1
- Python version used: 3.11

**Comment by ikonst** [comment]
We generally don't do anything sensible for loops - #14807, #14993, etc. We might want an issue label for those.

**Comment by eltoder** [comment]
I think #14807 is more ambitious because it needs a fixed point. #14993 might be the same issue, but the example there is more convoluted.

**Comment by ilevkivskyi** [comment]
This now gives an error on current master.
